pkg/federation/sdlmerge/sdlmerge.go

```diff
@@ -322,3 +325,66 @@ func (_ ScalarSharedType) valueRefs() []int {
 func (_ ScalarSharedType) valueName(_ int) string {
 	return ""
 }
+
+type PotentialEntityType interface {
+	getWalker() *astvisitor.Walker
+	getDocument() *ast.Document
+	assessValidEntity(ref int, nameBytes []byte) bool
+}
+
+func getPrimaryKeys(p PotentialEntityType, n *normalizer, name string, directiveRefs []int) map[string]bool {
```

Comment:
I propose to create a new type somewhat like that which will store entities info and will be responsible for checking their validity

```go
type EntitiesChecker {
 entities map[string]map[string]bool
}

func (checker *EntitiesChecker) getPrimaryKeys(document *ast.Document, name string, directiveRefs []int) (map[string]bool, error) {
}

func (checker *EntitiesChecker) checkAllPrimaryKeyReferencesAreExternal(document *ast.Document, name string, primaryKeys map[string]bool, fieldRefs []int) error {
}
```

This EntitiesChecker could be shared between visitors
```
type normalizer struct {
	walkers  []*astvisitor.Walker
	entityChecker *EntityChecker
}

newCollectValidEntitiesVisitor(m.entityChecker),
```

You could pass document instead of normalizer to `getPrimaryKeys` and `checkAllPrimaryKeyReferencesAreExternal` and return an error instead of calling walker stop from them, so each visitor will be responsive for stoping walker

smth like this 

```go
func (e extendObjectTypeDefinitionVisitor) assessValidEntity(ref int, nameBytes []byte) bool {
   // ...
  primaryKeys, err := e.entityChecker.getPrimaryKeys(e, e.normalizer, name, extension.Directives.Refs)
  if err != nil {
    e.Walker.StopWithExternalError(err)
    return
  }

  err = e.entityChecker.checkAllPrimaryKeyReferencesAreExternal(e, name, primaryKeys, extension.FieldsDefinition.Refs)
  if err != nil {
    e.Walker.StopWithExternalError(err)
    return

  }
}
```

In this way you will not have a need in a `type PotentialEntityType interface {`